File: leetcode_py/Subsets.py
class Solution:
    def helper(self, res, path, nums, idx):
        res.append(path)
        for i in range(idx, len(nums)):
            self.helper(res, path+[nums[i]], nums, i+1)

    # @param {integer[]} nums
    # @return {integer[][]}
    def subsets(self, nums):
        res = []
        self.helper(res, [], sorted(nums), 0)
        return res

class Solution:
    def helper(self, res, path, idx, nums):
        if idx == len(nums):
            res.append(path)
            return
        self.helper(res, path, idx+1, nums)
        self.helper(res, path+[nums[idx]], idx+1, nums)

    # @param {integer[]} nums
    # @return {integer[][]}
    def subsets(self, nums):
        res = []
        self.helper(res, [], 0, sorted(nums))
        return res

import datetime
# dp
class Solution:
    # @param {integer[]} nums
    # @return {integer[][]}
    def subsets(self, nums):
        nums.sort()
        res=[[]]
        for i in nums:
            tmp = res[:]
            for j in res:
                tmp.append(j + [i])
            res = tmp
        return res

File: leetcode_js/Subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var helper = function(nums, start, path, res){
    res.push(path);
    for(var i = start; i < nums.length; i++){
        helper(nums, i+1, path.concat(nums[i]), res);
    }
//    if(path.length > 0) res.push(path);
//    for(var i = start; i < nums.length; i++){
//        helper(nums, i+1, path, res);
//        helper(nums, i+1, path.concat(nums[i]), res);
//    }
}
var subsets = function(nums) {
    nums.sort(function(a,b){return a-b;})
    var res = [];
    helper(nums, 0, [], res);
    return res;
};